### PR TITLE
separate timer queue ack manager in separate file, add functionality …

### PR DESCRIPTION
**Pull request by wxing1292**
…to timer queue processor to be cluster aware

rename SequenceID -> TimerSequenceID

Standby timer processing logic && domain failover timer ack manager scan DB for domain specific timer in next PR.

### CODE REVIEW

File: service/history/timerQueueAckMgr.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+		shard         ShardContext
+		executionMgr  persistence.ExecutionManager
+		logger        bark.Logger
+		metricsClient metrics.Client
+		lastUpdated   time.Time
+		config        *Config
+
+		sync.Mutex
+		// outstanding tasks, key cluster name, value map of outstanding timer task -> finished (true)
+		clusterOutstandingTasks map[string]map[TimerSequenceID]bool
```
Comment (samarabbas) [inline service/history/timerQueueAckMgr.go]: I think we should keep this component simple by not making it aware of all clusters.  We should have a separate instance of TimerQueueProcessor with its own AckManager per cluster.

File: service/history/shardContext.go [hunk 2]
```diff
@@ -58,8 +58,8 @@ type (
 		GetConfig() *Config
 		GetLogger() bark.Logger
 		GetMetricsClient() metrics.Client
-		GetTimerAckLevel() time.Time
-		UpdateTimerAckLevel(ackLevel time.Time) error
+		GetTimerAckLevel(cluster string) time.Time
```
Comment (samarabbas) [inline service/history/shardContext.go]: Now we are going to have too many components managing ack levels separately.  Currently each update is a write to cassandra, this is not going to scale once cross dc support is in production.  Let's put a TODO to accept updates from individual components and have a separate job which updates all the ack levels in a single write for the entire shard.

File: service/history/shardContext.go [hunk 3]
```diff
@@ -121,6 +121,13 @@ func (s *shardContextImpl) GetTransferAckLevel() int64 {
 	s.RLock()
 	defer s.RUnlock()
 
+	// TODO change make this cluster input parameter
+	cluster := s.GetService().GetClusterMetadata().GetCurrentClusterName()
+	// if can find corresponding ack level in the cluster to timer ack level map
```
Comment (samarabbas) [inline service/history/shardContext.go]: nit: fix comment

File: service/history/timerQueueProcessor.go [hunk 4]
```diff
@@ -169,6 +156,12 @@ func (t *timerQueueProcessorImpl) NotifyNewTimers(timerTasks []persistence.Task)
 	}
 }
 
+// NotifyNewTimerEvents - Notify the processor about the new standby history events arrival.
+// This should be called each time new timer events arrives, otherwise timers maybe fired unexpected.
+func (t *timerQueueProcessorImpl) NotifyNewTimerEvents() {
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: I think we should have the same API for notifying new timers whether it is for active processor or standby processor.  Basically we should have a baseTimerQueueProcessor which has common logic and then activeTimerQueueProcessor and passiveTimerQueueProcessor with their own handling of TimerTasks.  But the way you notify both should be exactly the same.

File: service/history/shardContext.go (hunk 2 above)
Comment (wxing1292) [reply service/history/shardContext.go]: track in: https://github.com/uber/cadence/issues/634

File: service/history/shardContext.go (hunk 3 above)
Comment (wxing1292) [reply service/history/shardContext.go]: done

File: service/history/timerQueueProcessor.go (hunk 4 above)
Comment (wxing1292) [reply service/history/timerQueueProcessor.go]: this will be done in the next PR

**Pull request merged**